include/deal.II/matrix_free/dof_info.templates.h

```diff
@@ -283,13 +283,38 @@ namespace internal
       const TriaIterator<DoFCellAccessor<dim, dim, false>> &cell,
       std::vector<types::global_dof_index> &                dof_indices)
     {
-      const ArrayView<ConstraintKinds> mask_view(
-        hanging_node_constraint_masks.data() +
-          cell_number * cell->get_fe().n_components(),
-        cell->get_fe().n_components());
-
-      return hanging_nodes.setup_constraints(
-        cell, {}, lexicographic_mapping, dof_indices, mask_view);
+      // 1) check if finite elements support fast hanging-node algorithm
+      this->hanging_node_constraint_masks_comp =
+        hanging_nodes.compute_supported_components(
+          cell->get_dof_handler().get_fe_collection());
+
+      if ([](const auto &supported_components) {
+            return std::none_of(supported_components.begin(),
+                                supported_components.end(),
+                                [](const auto &a) {
+                                  return *std::max_element(a.begin(), a.end());
+                                });
+          }(hanging_node_constraint_masks_comp))
+        return false;
+
+      // 2) determine the refinement configuration of the cell
+      const auto refinement_configuration =
+        hanging_nodes.compute_refinement_configuration(cell);
```

Comment:
@kronbichler  We should be able to store the refinement configuration once per `MatrixFree` object. What would be the right place? `DoFInfo` sets up the information once per `DoFHandler`...

Reply:
We have https://github.com/dealii/dealii/blob/f9c375f53a798084a921ca47f8662b4ff38941cd/include/deal.II/matrix_free/matrix_free.h#L2107-L2119 which is a similar case as the one here, so we might do this in the generic `MatrixFree` class. Or we create a small struct that collects both the constraint weights and refinement configuration that get shared between different `DoFInfo` objects. I guess it depends if we anticipate additional arrays to be collected, as 3 objects is still somewhat usable.